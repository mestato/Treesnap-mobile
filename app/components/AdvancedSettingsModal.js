import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {
  Modal,
  View,
  StyleSheet,
  Text,
  Platform,
  KeyboardAvoidingView,
  TouchableOpacity
} from 'react-native'
import Elevation from '../helpers/Elevation'
import Colors from '../helpers/Colors'
import {isIphoneX} from 'react-native-iphone-x-helper'
import IonIcon from 'react-native-vector-icons/Ionicons'

const isAndroid = Platform.OS === 'android'

export default class AdvancedSettingsModal extends Component {
  constructor(props) {
    super(props)

    this.state = {
      isPrivate         : false,
      hasPrivateComments: false
    }
  }

  componentDidMount() {
    let {
          isPrivate,
          hasPrivateComments
        } = this.props.initialValues

    this.setState({
      isPrivate,
      hasPrivateComments
    })
  }

  render() {
    let commentsPrivacyCheckbox
    if (this.state.hasPrivateComments) {
      commentsPrivacyCheckbox = <IonIcon name={'md-checkbox'} size={20} color={Colors.primary}/>
    } else {
      commentsPrivacyCheckbox = <IonIcon name={'md-square-outline'} size={20} color={'#aaa'}/>
    }

    let isPrivateCheckbox
    if (this.state.isPrivate) {
      isPrivateCheckbox = <IonIcon name={'md-checkbox'} size={20} color={Colors.primary}/>
    } else {
      isPrivateCheckbox = <IonIcon name={'md-square-outline'} size={20} color={'#aaa'}/>
    }

    return (
      <Modal
        animationType="slide"
        visible={this.props.visible}
        onRequestClose={this.props.onRequestClose}
      >
        <View style={styles.modalHeader}>
          <Text style={styles.modalHeaderText}>
            Advanced Options
          </Text>
        </View>
        <KeyboardAvoidingView
          style={{flex: 1, backgroundColor: '#f4f4f4'}}
          {...(isAndroid ? null : {behavior: 'padding'})}>
          <View style={{flex: 1}}>

            <Text style={styles.title}>
              Privacy Settings
            </Text>
            <View style={styles.card}>
              <TouchableOpacity style={[styles.formGroup, {flex: 0, alignItems: 'center'}]}
                                activeOpacity={1}
                                onPress={() => {
                                  this.setState({
                                    hasPrivateComments: !this.state.hasPrivateComments
                                  }, () => {
                                    this.props.onChange(this.state)
                                  })
                                }}>
                <View style={{flex: 1, paddingVertical: 5, flexDirection: 'column'}}>
                  <Text style={[styles.label, {flex: 1}]}>
                    Mark comments as private
                  </Text>
                  <Text style={styles.helpText}>
                    {this.state.hasPrivateComments ?
                      'Visible only to you'
                      : 'Will be shared publicly'}
                  </Text>
                </View>
                <Text style={{
                  flex          : 0,
                  color         : this.state.hasPrivateComments ? '#444' : '#aaa',
                  width         : 40,
                  alignItems    : 'center',
                  justifyContent: 'center'
                }}>
                  {commentsPrivacyCheckbox}
                </Text>
              </TouchableOpacity>

              <TouchableOpacity style={[styles.formGroup, {flex: 0, alignItems: 'center'}]}
                                activeOpacity={1}
                                onPress={() => {
                                  this.setState({
                                    isPrivate: !this.state.isPrivate
                                  }, () => {
                                    this.props.onChange(this.state)
                                  })
                                }}>
                <View style={{flex: 1, paddingVertical: 5, flexDirection: 'column'}}>
                  <Text style={[styles.label, {flex: 1}]}>
                    Mark observation as private
                  </Text>
                  <Text style={styles.helpText}>
                    {this.state.isPrivate ?
                      'Visible only to you and our partners'
                      : 'Will be shared publicly'}
                  </Text>
                </View>
                <Text style={{
                  flex          : 0,
                  color         : this.state.isPrivate ? '#444' : '#aaa',
                  width         : 40,
                  alignItems    : 'center',
                  justifyContent: 'center'
                }}>
                  {isPrivateCheckbox}
                </Text>
              </TouchableOpacity>
            </View>
          </View>
          <View style={styles.footer}>
            <TouchableOpacity onPress={() => this.props.onRequestClose()}>
              <Text style={styles.doneText}>DONE</Text>
            </TouchableOpacity>
          </View>
        </KeyboardAvoidingView>
      </Modal>
    )
  }
}

AdvancedSettingsModal.propTypes = {
  visible       : PropTypes.bool.isRequired,
  onChange      : PropTypes.func.isRequired,
  onRequestClose: PropTypes.func.isRequired,
  initialValues : PropTypes.object.isRequired
}

AdvancedSettingsModal.defaultProps = {}

function getVerticalPadding() {
  if (Platform.OS === 'android') {
    return 0
  } else {
    if (isIphoneX()) {
      return 30
    }
    return 15
  }
}

const styles = StyleSheet.create({
  modalHeader: {
    backgroundColor: Colors.primary,
    paddingTop     : getVerticalPadding(),
    paddingBottom  : 10,
    ...(new Elevation(2))
  },

  modalHeaderText: {
    color          : Colors.primaryText,
    textAlign      : 'center',
    fontWeight     : 'normal',
    fontSize       : 16,
    paddingVertical: 5
  },

  title: {
    fontSize         : 14,
    color            : '#444',
    fontWeight       : 'bold',
    paddingHorizontal: 5,
    paddingTop       : 10,
    marginBottom     : 5
  },

  formGroup: {
    flex             : 0,
    flexDirection    : 'row',
    alignItems       : 'center',
    borderBottomWidth: 1,
    borderBottomColor: '#dedede',
    padding          : 5,
    minHeight        : 60
  },

  label: {
    flex      : 0,
    color     : '#444',
    fontWeight: 'bold'
  },

  card: {
    flex           : 0,
    borderTopWidth : 1,
    borderTopColor : '#dedede',
    backgroundColor: '#fff'
  },

  footer: {
    height           : isIphoneX() ? 60 : 40,
    borderTopWidth   : 1,
    borderTopColor   : '#dedede',
    backgroundColor  : '#eee',
    alignItems       : 'flex-end',
    justifyContent   : 'center',
    paddingHorizontal: 10,
    paddingBottom    : isIphoneX() ? 20 : 0
  },

  doneText: {
    fontSize  : 14,
    fontWeight: 'bold',
    color     : Colors.primary
  },

  helpText: {
    flex         : 1,
    color        : '#999',
    flexDirection: 'column',
    flexWrap     : 'wrap'
  }
})
