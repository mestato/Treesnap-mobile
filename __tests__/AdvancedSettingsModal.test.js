import React from 'react'
import AdvancedSettingsModal from '../app/components/AdvancedSettingsModal'
import renderer from 'react-test-renderer'

describe('it renders correctly', () => {
  it('should match an AdvancedSettingsModal snapshot', () => {
    const tree = renderer.create(<AdvancedSettingsModal visible={true}
                                                        onChange={jest.fn()}
                                                        onRequestClose={jest.fn()}
                                                        initialValues={{}}/>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
