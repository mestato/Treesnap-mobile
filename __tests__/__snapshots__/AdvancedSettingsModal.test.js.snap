// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly should match an AdvancedSettingsModal snapshot 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  visible={true}
>
  <View
    style={
      Object {
        "backgroundColor": "#2A9D8F",
        "paddingBottom": 10,
        "paddingTop": 15,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 1,
        },
        "shadowOpacity": 0.18,
        "shadowRadius": 0.9,
        "zIndex": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#ffffff",
          "fontSize": 16,
          "fontWeight": "normal",
          "paddingVertical": 5,
          "textAlign": "center",
        }
      }
    >
      Advanced Options
    </Text>
  </View>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#f4f4f4",
          "flex": 1,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#444",
            "fontSize": 14,
            "fontWeight": "bold",
            "marginBottom": 5,
            "paddingHorizontal": 5,
            "paddingTop": 10,
          }
        }
      >
        Privacy Settings
      </Text>
      <View
        style={
          Object {
            "backgroundColor": "#fff",
            "borderTopColor": "#dedede",
            "borderTopWidth": 1,
            "flex": 0,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#dedede",
              "borderBottomWidth": 1,
              "flex": 0,
              "flexDirection": "row",
              "minHeight": 60,
              "opacity": 1,
              "padding": 5,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "column",
                "paddingVertical": 5,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#444",
                    "flex": 0,
                    "fontWeight": "bold",
                  },
                  Object {
                    "flex": 1,
                  },
                ]
              }
            >
              Mark comments as private
            </Text>
            <Text
              style={
                Object {
                  "color": "#999",
                  "flex": 1,
                  "flexDirection": "column",
                  "flexWrap": "wrap",
                }
              }
            >
              Will be shared publicly
            </Text>
          </View>
          <Text
            style={
              Object {
                "alignItems": "center",
                "color": "#aaa",
                "flex": 0,
                "justifyContent": "center",
                "width": 40,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#aaa",
                    "fontSize": 20,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Ionicons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </Text>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#dedede",
              "borderBottomWidth": 1,
              "flex": 0,
              "flexDirection": "row",
              "minHeight": 60,
              "opacity": 1,
              "padding": 5,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "column",
                "paddingVertical": 5,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#444",
                    "flex": 0,
                    "fontWeight": "bold",
                  },
                  Object {
                    "flex": 1,
                  },
                ]
              }
            >
              Mark observation as private
            </Text>
            <Text
              style={
                Object {
                  "color": "#999",
                  "flex": 1,
                  "flexDirection": "column",
                  "flexWrap": "wrap",
                }
              }
            >
              Will be shared publicly
            </Text>
          </View>
          <Text
            style={
              Object {
                "alignItems": "center",
                "color": "#aaa",
                "flex": 0,
                "justifyContent": "center",
                "width": 40,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#aaa",
                    "fontSize": 20,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Ionicons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "backgroundColor": "#eee",
          "borderTopColor": "#dedede",
          "borderTopWidth": 1,
          "height": 40,
          "justifyContent": "center",
          "paddingBottom": 0,
          "paddingHorizontal": 10,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2A9D8F",
              "fontSize": 14,
              "fontWeight": "bold",
            }
          }
        >
          DONE
        </Text>
      </View>
    </View>
  </View>
</Modal>
`;
